# Recommendation Systems Project

• DOMAIN: Smartphone, Electronics

• CONTEXT: India is the second largest market globally for smartphones after China. About 134 million smartphones were sold across India in the year 2017 and is estimated to increase to about 442 million in 2022. India ranked second in the average time spent on mobile web by smartphone users across Asia Pacific. The combination of very high sales volumes and the average smartphone consumer behaviour has made India a very attractive market for foreign vendors. As per Consumer behaviour, 97% of consumers turn to a search engine when they are buying a product vs. 15% who turn to social media. If a seller succeeds to publish smartphones based on user’s behaviour/choice at the right place, there are 90% chances that user will enquire for the same. This Case Study is targeted to build a recommendation system 
based on individual consumer’s behaviour or choice.

• DATA DESCRIPTION: 

• author : name of the person who gave the rating

• country : country the person who gave the rating belongs to

• data : date of the rating

• domain: website from which the rating was taken from

• extract: rating content

• language: language in which the rating was given

• product: name of the product/mobile phone for which the rating was given

• score: average rating for the phone

• score_max: highest rating given for the phone

• source: source from where the rating was taken

• PROJECT OBJECTIVE: 

We will build a recommendation system using popularity based and collaborative filtering methods to recommend mobile phones to a user which are most popular and personalised respectively

## Q1 A - Merge all the provided CSVs into one data-frame.

In [85]:
import numpy as np
import pandas as pd

In [86]:
phdf_1 = pd.read_csv("phone_user_review_file_1.csv",encoding='latin1')
phdf_1.head()

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   

                                             extract       author  \
0  As a diehard Samsung fan who has had every Sam...   CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...    james0923   
2  Adequate feel. Nice heft. Processor's still sl...     R. Craig   
3  Never disappointed. One of the reasons I've be...  Buster2020    
4  I've now found that i'm in a group of people t...   S Ate Mine   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8

In [87]:
phdf_2 = pd.read_csv("phone_user_review_file_2.csv",encoding='latin1')
phdf_2.head()

phone_url       date lang country  source        domain  \
0  /cellphones/leagoo-lead-7/  4/15/2015   en      us  Amazon    amazon.com   
1  /cellphones/leagoo-lead-7/  5/23/2015   en      gb  Amazon  amazon.co.uk   
2  /cellphones/leagoo-lead-7/  4/27/2015   en      gb  Amazon  amazon.co.uk   
3  /cellphones/leagoo-lead-7/  4/22/2015   en      gb  Amazon  amazon.co.uk   
4  /cellphones/leagoo-lead-7/  4/18/2015   en      gb  Amazon  amazon.co.uk   

   score  score_max                                            extract  \
0    2.0       10.0  The telephone headset is of poor quality , not...   
1   10.0       10.0  This is my first smartphone so I have nothing ...   
2    8.0       10.0  Great phone. Battery life not great but seems ...   
3   10.0       10.0      Best 90 quid I've ever spent on a smart phone   
4   10.0       10.0  I m happy with this phone.it s very good.thx team   

          author                                            product  
0           luis  Leagoo Lead7 5.0 Inch HD JDI LTPS Screen 3G Sm...  
1     Mark Lavin  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
2         tracey  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
3  Reuben Ingram  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
4         viorel  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...

In [88]:
phdf_3 = pd.read_csv("phone_user_review_file_3.csv",encoding='latin1')
phdf_3.head()

phone_url       date lang country  \
0  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  11/7/2015   pt      br   
1  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  10/2/2015   pt      br   
2  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt      br   
3  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt      br   
4  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/1/2015   pt      br   

      source            domain  score  score_max  \
0  Submarino  submarino.com.br    6.0       10.0   
1  Submarino  submarino.com.br   10.0       10.0   
2  Submarino  submarino.com.br   10.0       10.0   
3  Submarino  submarino.com.br    8.0       10.0   
4    Colombo    colombo.com.br    8.0       10.0   

                                             extract  \
0  recomendo, eu comprei um, a um ano, e agora co...   
1  Comprei um pouco desconfiada do site e do celu...   
2  Muito bom o produto, obvio que tem versões me...   
3  Unica ressalva fica para a camera que poderia ...   
4  Rapidez e atenção na entrega. O aparelho é ...   

                           author  \
0                herlington tesch   
1             Luisa Silva Marieta   
2                           Cyrus   
3       Marcela Santa Clara Brito   
4  Claudine Maria Kuhn Walendorff   

                                             product  
0  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
1  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
2  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
3  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
4  Smartphone Samsung Galaxy S3 Slim, Dual Chip, ...

In [89]:
phdf_4 = pd.read_csv("phone_user_review_file_4.csv",encoding='latin1')
phdf_4.head()

phone_url        date lang country  \
0  /cellphones/samsung-s7262-duos-galaxy-ace/   3/11/2015   en      us   
1  /cellphones/samsung-s7262-duos-galaxy-ace/  17/11/2015   en      in   
2  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
3  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
4  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   

   source      domain  score  score_max  \
0  Amazon  amazon.com    2.0       10.0   
1  Zopper  zopper.com   10.0       10.0   
2  Amazon   amazon.in    4.0       10.0   
3  Amazon   amazon.in    6.0       10.0   
4  Amazon   amazon.in   10.0       10.0   

                                             extract            author  \
0  was not conpatable with my phone as stated. I ...  Frances DeSimone   
1  Decent Functions and Easy to Operate Pros:- Th...     Expert Review   
2  Not Good Phone such price. Hang too much and v...   Amazon Customer   
3                               not bad for features   Amazon Customer   
4                                  Excellent product               NHK   

                                             product  
0  Samsung Galaxy Star Pro DUOS S7262 Unlocked Ce...  
1                Samsung Galaxy Star Pro S7262 Black  
2  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
3  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
4  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)

In [90]:
phdf_5 = pd.read_csv("phone_user_review_file_5.csv",encoding='latin1')
phdf_5.head()

phone_url       date lang country      source  \
0  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
1  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
2  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
3  /cellphones/karbonn-k1616/  4/25/2014   en      in     Naaptol   
4  /cellphones/karbonn-k1616/  4/23/2013   en      in     Naaptol   

          domain  score  score_max  \
0  91mobiles.com    2.0       10.0   
1  91mobiles.com    6.0       10.0   
2  91mobiles.com    4.0       10.0   
3    naaptol.com   10.0       10.0   
4    naaptol.com   10.0       10.0   

                                             extract           author  \
0  I bought 1 month before. currently speaker is ...        venkatesh   
1  I just bought one week back, I have Airtel con...           Venkat   
2  one problem in this handset opera is not worki...           krrish   
3  here Karbonn comes up with an another excellen...  BRIJESH CHAUHAN   
4  What a phone, all so on Naaptol my god 23% off...    Suraj CHAUHAN   

                 product  
0          Karbonn K1616  
1          Karbonn K1616  
2          Karbonn K1616  
3  Karbonn K1616 - Black  
4  Karbonn K1616 - Black

In [91]:
phdf_6 = pd.read_csv("phone_user_review_file_6.csv",encoding='latin1')
phdf_6.head()

phone_url        date lang country  \
0  /cellphones/samsung-instinct-sph-m800/   9/16/2011   en      us   
1  /cellphones/samsung-instinct-sph-m800/   2/13/2014   en      us   
2  /cellphones/samsung-instinct-sph-m800/  12/30/2011   en      us   
3  /cellphones/samsung-instinct-sph-m800/  10/18/2008   en      us   
4  /cellphones/samsung-instinct-sph-m800/    9/6/2008   en      us   

          source             domain  score  score_max  \
0    Phone Arena     phonearena.com    8.0       10.0   
1         Amazon         amazon.com    6.0       10.0   
2    Phone Scoop     phonescoop.com    9.0       10.0   
3  HandCellPhone  handcellphone.com    4.0       10.0   
4   Reviewed.com       reviewed.com    6.0       10.0   

                                             extract      author  \
0  I've had the phone for awhile and it's a prett...  ajabrams95   
1  to be clear it is not the sellers fault that t...   Stephanie   
2  Well i love this phone. i have had ton of phon...    snickers   
3  I have had my Instinct for several months now ...         A4C   
4  i have had this instinct phone for about two m...   betaBgood   

                     product  
0        Samsung Instinct HD  
1  Samsung SPH M800 Instinct  
2              Instinct M800  
3           Samsung Instinct  
4           Samsung Instinct

In [92]:
print(phdf_1.shape)
print(phdf_2.shape)
print(phdf_3.shape)
print(phdf_4.shape)
print(phdf_5.shape)
print(phdf_6.shape)

(374910, 11)
(114925, 11)
(312961, 11)
(98284, 11)
(350216, 11)
(163837, 11)


In [93]:
from functools import reduce

In [94]:
dfs = [phdf_1,phdf_2,phdf_3,phdf_4,phdf_5,phdf_6]

In [95]:
df_1 = reduce(lambda  left,right: pd.merge(left,right,how='outer'), dfs)
df_1.shape

(1414877, 11)

In [96]:
df_1.tail()

phone_url       date lang country source  \
1414872  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   Ciao   
1414873  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   Ciao   
1414874  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   Ciao   
1414875  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   Ciao   
1414876  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   Ciao   

          domain  score  score_max  \
1414872  ciao.de    2.0       10.0   
1414873  ciao.de   10.0       10.0   
1414874  ciao.de    2.0       10.0   
1414875  ciao.de    8.0       10.0   
1414876  ciao.de    2.0       10.0   

                                                   extract        author  \
1414872  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
1414873  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
1414874  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
1414875  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
1414876  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                         product  
1414872  Alcatel Club Plus Handy  
1414873  Alcatel Club Plus Handy  
1414874  Alcatel Club Plus Handy  
1414875  Alcatel Club Plus Handy  
1414876  Alcatel Club Plus Handy

In [97]:
df_1.head()

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   

                                             extract       author  \
0  As a diehard Samsung fan who has had every Sam...   CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...    james0923   
2  Adequate feel. Nice heft. Processor's still sl...     R. Craig   
3  Never disappointed. One of the reasons I've be...  Buster2020    
4  I've now found that i'm in a group of people t...   S Ate Mine   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8

## Q1 B - Explore, understand the Data and share at least 2 observations.

In [98]:
df_1.shape

(1414877, 11)

- There is 11 columns and 1414877 rows are there in the Dataframe

In [99]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1414877 entries, 0 to 1414876
Data columns (total 11 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   phone_url  1414877 non-null  object 
 1   date       1414877 non-null  object 
 2   lang       1414877 non-null  object 
 3   country    1414877 non-null  object 
 4   source     1414877 non-null  object 
 5   domain     1414877 non-null  object 
 6   score      1351388 non-null  float64
 7   score_max  1351388 non-null  float64
 8   extract    1395518 non-null  object 
 9   author     1351675 non-null  object 
 10  product    1414876 non-null  object 
dtypes: float64(2), object(9)
memory usage: 129.5+ MB


- Except score and score_max columns are numerical type(float) where remaining are object type
- It seems to be there is some null values exists in columns

In [100]:
df_1.describe()

score  score_max
count  1.351388e+06  1351388.0
mean   8.007118e+00       10.0
std    2.616063e+00        0.0
min    2.000000e-01       10.0
25%    7.200000e+00       10.0
50%    9.200000e+00       10.0
75%    1.000000e+01       10.0
max    1.000000e+01       10.0

- The statistical information as
   - the score column having the min rating of 2 and max rating of 10
   - the score_max column having only one value i.e only 10

In [101]:
df_1.dtypes

phone_url     object
date          object
lang          object
country       object
source        object
domain        object
score        float64
score_max    float64
extract       object
author        object
product       object
dtype: object

## Q1 C - Round off scores to the nearest integers.

In [102]:
df_1['score'] = round(df_1.score)

In [103]:
df_1.tail()

phone_url       date lang country source  \
1414872  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   Ciao   
1414873  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   Ciao   
1414874  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   Ciao   
1414875  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   Ciao   
1414876  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   Ciao   

          domain  score  score_max  \
1414872  ciao.de    2.0       10.0   
1414873  ciao.de   10.0       10.0   
1414874  ciao.de    2.0       10.0   
1414875  ciao.de    8.0       10.0   
1414876  ciao.de    2.0       10.0   

                                                   extract        author  \
1414872  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
1414873  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
1414874  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
1414875  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
1414876  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                         product  
1414872  Alcatel Club Plus Handy  
1414873  Alcatel Club Plus Handy  
1414874  Alcatel Club Plus Handy  
1414875  Alcatel Club Plus Handy  
1414876  Alcatel Club Plus Handy

In [104]:
df_1['score'] = df_1['score'].astype('Int64')

In [105]:
df_1.tail()

phone_url       date lang country source  \
1414872  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   Ciao   
1414873  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   Ciao   
1414874  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   Ciao   
1414875  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   Ciao   
1414876  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   Ciao   

          domain  score  score_max  \
1414872  ciao.de      2       10.0   
1414873  ciao.de     10       10.0   
1414874  ciao.de      2       10.0   
1414875  ciao.de      8       10.0   
1414876  ciao.de      2       10.0   

                                                   extract        author  \
1414872  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
1414873  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
1414874  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
1414875  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
1414876  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                         product  
1414872  Alcatel Club Plus Handy  
1414873  Alcatel Club Plus Handy  
1414874  Alcatel Club Plus Handy  
1414875  Alcatel Club Plus Handy  
1414876  Alcatel Club Plus Handy

In [106]:
df_1.dtypes

phone_url     object
date          object
lang          object
country       object
source        object
domain        object
score          Int64
score_max    float64
extract       object
author        object
product       object
dtype: object

## Q1 D - Check for missing values. Impute the missing values, if any.

In [107]:
# checking Null values percentage for each column
(df_1.isnull().sum()/len(df_1))*100

phone_url    0.000000
date         0.000000
lang         0.000000
country      0.000000
source       0.000000
domain       0.000000
score        4.487245
score_max    4.487245
extract      1.368246
author       4.466961
product      0.000071
dtype: float64

- It seems to be there is score and score_max and author columns having nearly 4.5% of missing values

In [108]:
df_1.isnull().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        63489
score_max    63489
extract      19359
author       63202
product          1
dtype: int64

- By the above observation of Null values in each column we may notice that score and score_max having similar missing values
  they rows that these two columns had missed same rows

- So I am dropping those rows

In [109]:
df_2 = df_1.dropna()
df_2.isnull().sum()

phone_url    0
date         0
lang         0
country      0
source       0
domain       0
score        0
score_max    0
extract      0
author       0
product      0
dtype: int64

## Q1 E - Check for duplicate values and remove them, if any.

In [110]:
df_2.duplicated().sum()

4226

- There are some dublicates also present we remove them

In [111]:
df_2[df_2.duplicated()]

phone_url        date lang country  \
342               /cellphones/samsung-galaxy-s8/   4/26/2017   nl      nl   
343               /cellphones/samsung-galaxy-s8/   4/26/2017   nl      nl   
371               /cellphones/samsung-galaxy-s8/   4/25/2017   nl      nl   
373               /cellphones/samsung-galaxy-s8/   4/25/2017   nl      nl   
3718     /cellphones/samsung-galaxy-s6-edgeplus/   10/4/2015   en      us   
...                                          ...         ...  ...     ...   
1394842                 /cellphones/nokia-6310i/    5/3/2014   ru      ru   
1394953                 /cellphones/nokia-6310i/   7/22/2011   ru      ru   
1394999                 /cellphones/nokia-6310i/  11/19/2009   ru      ru   
1397864                  /cellphones/nokia-8850/   8/23/2014   ru      ru   
1411204          /cellphones/philips-genie-2000/   8/20/2000   it      it   

           source            domain  score  score_max  \
342      Coolblue       coolblue.nl     10       10.0   
343      Coolblue       coolblue.nl     10       10.0   
371      Coolblue       coolblue.nl     10       10.0   
373      Coolblue       coolblue.nl     10       10.0   
3718       Amazon        amazon.com     10       10.0   
...           ...               ...    ...        ...   
1394842    Yandex  market.yandex.ru     10       10.0   
1394953    Yandex  market.yandex.ru     10       10.0   
1394999    Yandex  market.yandex.ru     10       10.0   
1397864    Yandex  market.yandex.ru     10       10.0   
1411204      Ciao           ciao.it      8       10.0   

                                                   extract  \
342      Ik vind het tot nu toe een heel erg aangenaam ...   
343      Ik vind het tot nu toe een heel erg aangenaam ...   
371      Tot nu toe in de eerste dagen weinig tot geen ...   
373                                   Wat een huge scherm!   
3718     For readers, my opinions on this device (ATT 3...   
...                                                    ...   
1394842  Телефоном пользовался да...   
1394953  Очень жалею об утрате это...   
1394999  У меня "NOKIA 6310i" уже более 3-...   
1397864  Данный отзыв написан лишÑ...   
1411204  Il Genie e' un buon telefono, scorrimente del ...   

                                author  \
342                  Stefanie Hinneman   
343                  Stefanie Hinneman   
371                     Robin Winckens   
373                            B Yapar   
3718                       FrozenCloud   
...                                ...   
1394842  Бикбаев Айбулат   
1394953                  alienwarfiend   
1394999                      Yurigin70   
1397864        Иванов Денис   
1411204                      tonyblack   

                                                   product  
342                                Samsung Galaxy S8 Zwart  
343                                Samsung Galaxy S8 Zwart  
371                                Samsung Galaxy S8 Zwart  
373                                Samsung Galaxy S8 Zwart  
3718     Samsung Galaxy S6 Edge+, Black 64GB (Verizon W...  
...                                                    ...  
1394842                                        Nokia 6310i  
1394953                                        Nokia 6310i  
1394999                                        Nokia 6310i  
1397864                                         Nokia 8850  
1411204                                 Philips Genie 2000  

[4226 rows x 11 columns]

In [112]:
df_2 = df_2.drop_duplicates()

In [113]:
df_2.duplicated().sum()

0

## Q1 F - Keep only 1 Million data samples. Use random state=612.

In [114]:
df_3 = df_2.sample(n=1000000,random_state=612)
print(df_3.shape)

(1000000, 11)


## Q1 G - Drop irrelevant features. Keep features like Author, Product, and Score.

In [115]:
print("the columns in the dataframe have : \n " ,df_2.columns)

the columns in the dataframe have : 
  Index(['phone_url', 'date', 'lang', 'country', 'source', 'domain', 'score',
       'score_max', 'extract', 'author', 'product'],
      dtype='object')


In [116]:
for x in df_2.columns:
    a = df_2[x].unique()
    print(len(a))
    print(x  , (len(a)/len(df_2))*100)

5508
phone_url 0.433210611300442
7715
date 0.6067937302438108
20
lang 0.0015730232799580316
40
country 0.003146046559916063
284
source 0.02233693057540405
334
domain 0.02626948877529913
11
score 0.0008651628039769174
1
score_max 7.865116399790159e-05
1219926
extract 95.94859989130408
770521
author 60.60237353482712
54872
product 4.315746670892856


In [117]:
df_4 = df_2.drop(['score_max','phone_url','extract'],axis=1)
df_4.columns

Index(['date', 'lang', 'country', 'source', 'domain', 'score', 'author',
       'product'],
      dtype='object')

In [118]:
df_5 = df_4.drop(['date','lang','country','source','domain'],axis = 1)

## Q2 A - Identify the most rated features.

In [119]:
df_5.groupby('product')['score'].count().sort_values(ascending=False)

product
Lenovo Vibe K4 Note (White,16GB)                                     5223
Lenovo Vibe K4 Note (Black, 16GB)                                    4389
OnePlus 3 (Graphite, 64 GB)                                          4103
OnePlus 3 (Soft Gold, 64 GB)                                         3557
Huawei P8lite zwart / 16 GB                                          2707
                                                                     ... 
Nokia LUMIA 530 VERT                                                    1
Nokia LUMIA 530 BLANC                                                   1
Nokia KIT PIETON ORIGIN STEREO NOKIA WH-102 5800/E52/C5 noir bulk       1
Nokia Illuvial 6700 Rosa                                                1
摩托罗拉 C168i                                                      1
Name: score, Length: 54872, dtype: int64

In [120]:
df_5.groupby('product')['score'].value_counts().to_frame()

score
product                                            score       
'Smartphone Meizu Pro 5, 5,7 pouces avec Exynos... 8          1
'Sony Xperia X (F5122) – White – Dual Sim (... 10         1
'Sony Xperia X (F5122) – rosa – Dual Sim (G... 10         1
(7.62 cm (3 )Afficheur/écran, 2 MPixCaméra;bl... 6          1
(CUBOT) GT88 5.5" qHD 1.3GHz MTK6572 2-Core And... 8          1
...                                                         ...
ساÙ
سونج جالكسي J1 SM-J120FD بشر... 6          1
ساÙ
سونج جالكسي J1 SM-J120FD بشر... 10         3
                                                   8          2
هاواوي P9 بلوس - 64 جيجابايتØ... 10         1
摩托罗拉 C168i                                 10         1

[161016 rows x 1 columns]

## Q2 B - Identify the users with most number of reviews.

In [121]:
df_5.groupby('author')['score'].count().sort_values(ascending=False).to_frame()

score
author                                                   
Amazon Customer                                     76933
Cliente Amazon                                      19300
e-bit                                                8411
Client d'Amazon                                      7716
Amazon Kunde                                         4746
...                                                   ...
Luckencia Orphee                                        1
Lucki                                                   1
Luckie                                                  1
Luckie332                                               1
��������� X�� ���ï¿...      1

[770521 rows x 1 columns]

## Q2 C - Select the data with products having more than 50 ratings and users who have given more than 50 ratings. Report the shape of the final dataset.

In [122]:
df_6 = df_5.copy()

In [123]:
counts1 = pd.value_counts(df_6['author'])

In [124]:
counts1

Amazon Customer     76933
Cliente Amazon      19300
e-bit                8411
Client d'Amazon      7716
Amazon Kunde         4746
                    ...  
Houdini Oxendine        1
Desay Chbyani           1
bradwood135             1
MrsLisaPalmer           1
claudia0815             1
Name: author, Length: 770521, dtype: int64

In [125]:
df_6['author'].value_counts()

Amazon Customer     76933
Cliente Amazon      19300
e-bit                8411
Client d'Amazon      7716
Amazon Kunde         4746
                    ...  
Houdini Oxendine        1
Desay Chbyani           1
bradwood135             1
MrsLisaPalmer           1
claudia0815             1
Name: author, Length: 770521, dtype: int64

In [126]:
df_6 = df_6[df_6['author'].isin(counts1[counts1 >= 50].index)]

In [127]:
df_6.shape

(233425, 3)

In [128]:
df_6['author'].value_counts()

Amazon Customer                76933
Cliente Amazon                 19300
e-bit                           8411
Client d'Amazon                 7716
Amazon Kunde                    4746
                               ...  
???????????????? ??????????       50
Gerard                            50
Guy                               50
Sander                            50
Sachin                            50
Name: author, Length: 831, dtype: int64

In [129]:
counts2 = pd.value_counts(df_6['product'])

In [130]:
df_6 = df_6[df_6['product'].isin(counts2[counts2 >= 50].index)]

In [131]:
df_6['product'].value_counts()

Lenovo Vibe K4 Note (White,16GB)                                                                                                                    3136
Lenovo Vibe K4 Note (Black, 16GB)                                                                                                                   2613
OnePlus 3 (Graphite, 64 GB)                                                                                                                         1910
OnePlus 3 (Soft Gold, 64 GB)                                                                                                                        1817
Lenovo Vibe K5 (Gold, VoLTE update)                                                                                                                 1609
                                                                                                                                                    ... 
BlackBerry Z3 (Black, 8GB)                                                        

In [132]:
final_df = df_6.copy()

## Q3 - Build a popularity based model and recommend top 5 mobile phones.

In [133]:
ppt_data_grouped = final_df.groupby('product').agg({'score': 'count'}).reset_index()
ppt_data_grouped.head()

product  score
0  AICEK Coque Samsung Galaxy A3 2016, AICEK Etui...     91
1  ASUS ZenFone 2 Laser Unlocked Smartphone, 3GB ...     80
2  ASUS ZenFone 2 Unlocked Smartphone, 16GB, Silv...     66
3  Access-Discount SAMSUNG GALAXY J3 2016 coque s...     76
4  Alcatel One Touch 20-04G Telefono Cellulare, Nero    234

In [134]:
#Sort the products on recommendation score 
ppt_data_sort = ppt_data_grouped.sort_values(['score', 'product'], ascending = [0,1]) 
      
#Generate a recommendation rank based upon score 
ppt_data_sort['Rank'] = ppt_data_sort['score'].rank(ascending=0, method='first') 
          
#Get the top 5 recommendations 
popularity_recommendations = ppt_data_sort.head(5) 
popularity_recommendations 

product  score  Rank
257     Lenovo Vibe K4 Note (White,16GB)   3136   1.0
256    Lenovo Vibe K4 Note (Black, 16GB)   2613   2.0
347          OnePlus 3 (Graphite, 64 GB)   1910   3.0
348         OnePlus 3 (Soft Gold, 64 GB)   1817   4.0
258  Lenovo Vibe K5 (Gold, VoLTE update)   1609   5.0

In [135]:
final_df['product'].value_counts()

Lenovo Vibe K4 Note (White,16GB)                                                                                                                    3136
Lenovo Vibe K4 Note (Black, 16GB)                                                                                                                   2613
OnePlus 3 (Graphite, 64 GB)                                                                                                                         1910
OnePlus 3 (Soft Gold, 64 GB)                                                                                                                        1817
Lenovo Vibe K5 (Gold, VoLTE update)                                                                                                                 1609
                                                                                                                                                    ... 
BlackBerry Z3 (Black, 8GB)                                                        

## Q4 - Build a collaborative filtering model using SVD. You can use SVD from surprise or build it from scratch.

## Q5 - Evaluate the collaborative model. Print RMSE value.

In [136]:
import surprise
from surprise import Dataset,Reader

In [137]:
reader = Reader(rating_scale=(1, 10))
final_data = Dataset.load_from_df(final_df[[ 'author', 'product','score']], reader)

In [138]:
# Split data to train and test

from surprise.model_selection import train_test_split
trainset, testset = train_test_split(final_data, test_size=.20,random_state=612)

In [139]:
trainset.all_ratings()

<generator object Trainset.all_ratings at 0x000001F71E36E970>

In [140]:
print(trainset.to_raw_uid(6))

Дмитрий


In [141]:
print(testset[0])

('Andrea', 'Samsung Galaxy Note 3 Smartphone (14,5 cm (5,7 Zoll) AMOLED-Touchscreen, 2,3GHz, Quad-Core, 3GB RAM, 13 Megapixel Kamera, Android 4.3) weiÃ\x9f/gold', 10.0)


In [142]:
from surprise import SVD, KNNWithMeans
from surprise import accuracy

In [143]:
svd_model = SVD(n_factors=5,biased=False,random_state=612)
svd_model.fit(trainset)

In [144]:
test_pred = svd_model.test(testset)

In [145]:
# get RMSE
print("SVD Model : Test Set")
print(accuracy.rmse(test_pred))

SVD Model : Test Set
RMSE: 2.6858
2.68575860742736


### Userbased and Item based collabarative filtering

#### KNNWithMeans

#### user based

In [146]:
algo_user = KNNWithMeans(k=70, sim_options={ 'user_based': True})

algo_user.fit(trainset)

test_pred_u=algo_user.test(testset)

Computing the msd similarity matrix...
Done computing similarity matrix.


In [147]:
# get RMSE
print("User-based Model : Test Set")
accuracy.rmse(test_pred_u, verbose=True)

User-based Model : Test Set
RMSE: 2.6362


2.636150504065585

#### Item based

In [148]:
algo_item = KNNWithMeans(k=70, sim_options={ 'user_based': False})

algo_item.fit(trainset)

test_pred_item=algo_item.test(testset)

Computing the msd similarity matrix...
Done computing similarity matrix.


In [149]:
# get RMSE
print("Item-based Model : Test Set")
print(accuracy.rmse(test_pred_item,verbose=True))

Item-based Model : Test Set
RMSE: 2.6296
2.6296280749142054


- By comparing the all the RMSE scores item-item collabarative filtering gives lower value

## Q6 - Predict score (average rating) for test users.

In [150]:
test_pred_item[0]

Prediction(uid='Andrea', iid='Samsung Galaxy Note 3 Smartphone (14,5 cm (5,7 Zoll) AMOLED-Touchscreen, 2,3GHz, Quad-Core, 3GB RAM, 13 Megapixel Kamera, Android 4.3) weiÃ\x9f/gold', r_ui=10.0, est=7.503746746993757, details={'actual_k': 70, 'was_impossible': False})

In [151]:
test_pred_item[0][2],test_pred_item[0][4]

(10.0, {'actual_k': 70, 'was_impossible': False})

- the predicted score is the average rating of the nearest neighbor(similar item ) 

In [152]:
test_pred_item

[Prediction(uid='Andrea', iid='Samsung Galaxy Note 3 Smartphone (14,5 cm (5,7 Zoll) AMOLED-Touchscreen, 2,3GHz, Quad-Core, 3GB RAM, 13 Megapixel Kamera, Android 4.3) weiÃ\x9f/gold', r_ui=10.0, est=7.503746746993757, details={'actual_k': 70, 'was_impossible': False}),
 Prediction(uid='Amazon Customer', iid='OnePlus One (Sandstone Black, 64GB)', r_ui=8.0, est=6.828571428571428, details={'actual_k': 70, 'was_impossible': False}),
 Prediction(uid='Amazon Customer', iid='OnePlus X (Limited Edition Ceramic)', r_ui=6.0, est=6.942857142857143, details={'actual_k': 70, 'was_impossible': False}),
 Prediction(uid='Amazon Customer', iid='YU Yureka (Moondust Grey,16 GB)', r_ui=8.0, est=6.171428571428571, details={'actual_k': 70, 'was_impossible': False}),
 Prediction(uid='Patrick', iid='Honor 7 Smartphone débloqué 4G (Ecran: 5,2 pouces - 16 Go - Double Nano SIM - Android 5.0 Lollipop) Argent/Blanc', r_ui=8.0, est=8.415801426504578, details={'actual_k': 42, 'was_impossible': False}),
 Prediction(u

## Q7 - Report your findings and inferences.

- Most of the products having actual k value is 70 and mean ratings are given by them at the same time some of neighbors are less than 70
- almost three alogroithams namely SVD,user based collabartive and item based collabarative gives me similar RMSE values
     - SVD -- 2.69
     - USER based CF -- 2.64
     - Item based CF --- 2.62
- among all those alogorithams item based alogoritham performs well in case RMSE

## Q8 - Try and recommend top 5 products for test users.

In [153]:
testset[10]

('Francesco',
 'Lenovo Motorola Moto G 4G (2 Generazione) Smartphone, Display 5 Pollici, LTE, Fotocamera 8 MP, Memoria 8 GB, Android 5 Lollipop, Nero [Italia]',
 10.0)

- The Recommended products for Amazon user are :

In [154]:
pred = pd.DataFrame(test_pred_item)
pred[pred['uid'] == 'Amazon Customer'][['iid', 'r_ui','est']].sort_values(by = 'r_ui',ascending = False).head(5)

iid  r_ui       est
13913  OnePlus 3T (Soft Gold, 6GB RAM + 64GB memory)  10.0  8.085714
12788                              Nokia 105 (Black)  10.0  6.342857
12784                     BlackBerry Z3 (Black, 8GB)  10.0  5.524108
12781               Lenovo Vibe K4 Note (White,16GB)  10.0  6.371429
5822    OnePlus 3T (Gunmetal, 6GB RAM + 64GB memory)  10.0  8.514286

In [155]:
testset[50]

('Cliente Amazon', 'WIKO Fever 4G Smartphone, 16 GB, Dual SIM, Bianco', 10.0)

In [156]:
pred = pd.DataFrame(test_pred_item)
pred[pred['uid'] == 'Cliente Amazon'][['iid', 'r_ui','est']].sort_values(by = 'r_ui',ascending = False).head(5)

iid  r_ui       est
29     Samsung J320 Galaxy J3 (2016) Dual SIM 5", Qua...  10.0  8.377733
10783  Meizu M2 Note Smartphone, 5.5" Full HD, 4G, 13...  10.0  8.549070
10449  Huawei P9 Lite Smartphone, LTE, Display 5.2'' ...  10.0  9.314286
10515  Honor 7 Smartphone 4G, Display Full HD 5.2 Pol...  10.0  9.188836
10525  Huawei P8 lite Smartphone, Display 5.0" IPS, D...  10.0  8.171429

## Q9 - Try other techniques (Example: cross validation) to get better results.

In [157]:
algo_item1 = KNNWithMeans(k=30, sim_options={ 'user_based': False})

algo_item1.fit(trainset)


Computing the msd similarity matrix...
Done computing similarity matrix.


In [158]:
from surprise.model_selection import cross_validate
cross_validate(algo_item1, final_data, measures=['RMSE', 'MAE'], cv=3, verbose=True)

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE, MAE of algorithm KNNWithMeans on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    2.6347  2.6585  2.6648  2.6527  0.0130  
MAE (testset)     2.0676  2.0791  2.0857  2.0775  0.0075  
Fit time          39.44   40.21   48.55   42.73   4.13    
Test time         199.17  222.82  223.44  215.14  11.30   


{'test_rmse': array([2.63467812, 2.65849877, 2.66478094]),
 'test_mae': array([2.06761132, 2.07909168, 2.08566385]),
 'fit_time': (39.43509769439697, 40.21455240249634, 48.55276083946228),
 'test_time': (199.17112231254578, 222.8239631652832, 223.4365005493164)}

## Q10 - In what business scenario you should use popularity based Recommendation Systems ?

- Incase of we don't have any userinformation and for startups primarily popularity based recommendation systems are useful
- In the above scenario If a new customer arrive in ,we don't have any information about the user , In this situation we can recommend the popular ones
- for example if we taken any news paper online websites ,first they will show you popular ones
- for example if we see in amazon, if a new customer arrives in ,he searched for an item, it will show you first the most popular ones in their category

##  Q11 - In what business scenario you should use CF based Recommendation Systems ?

- Incase of we ae having large number users and millions of items CF based Recommendation systems will be useful
- In case of NETFLIX and Amazon there is huge million of users and items will be available, when a user given a rating for an item
   - we can show which items are similar to that item
   - we can analyze his buying pattern or intrest and try to find his neighbors and recommend those type of items what his neighbors intrested in
   - by using his intrests or ratings we can recommend items to other similar user
   - in case of customers who not rated some items , we can rate them by CF recommedation systems

## Q12 - What other possible methods can you think of which can further improve the recommendation for different users ?

- By using Hybrid recommender systems we can be able to improve the recommendation for different users
- Content based models can be used to solve the Cold start and Grey sheep problems in Collaborative filtering.